---
title: "Practical: Alevin combined"
subtitle: "Transcriptome Analysis Workshop"
author: "Adam Cribbs"
date: "24/01/2022"
output:
  html_document:
    theme: cosmo
    toc: yes
---

```{r, out.width = "40%", echo=FALSE}
htmltools::img(src = knitr::image_uri(file.path("logo.png")), 
               alt = 'logo', 
               style = 'position:absolute; top:0; right:0; padding:10px;',
               width='300')

```


My main reason for using Seurat over other single-cell software is that it contains an implimentation of Cononical Correlation Alignment (CCA) that allows different datasets to be merged.

In this Rmarkdown we will alignt two sets of data, 1000 PBMC cells that were ran using a version 3 10X kit and one that was ran using the v2 kit.

# Import the packages

```{r, include=FALSE}
knitr::opts_chunk$set(cache=FALSE, warning = FALSE, message = FALSE)
Sys.setenv(KMP_DUPLICATE_LIB_OK=TRUE)
library(Seurat)
library(tximport)
library(cowplot)
library(org.Hs.eg.db)
library(dplyr)
library("AnnotationDbi")
library(EnsDb.Hsapiens.v75)
library(DT)
```

# Import the data 

The experiment was analysed using [cgat-developer](https://github.com/cgat-developers/cgat-core) pipelines for quality of sequencing. Read quality looked good. 

Next I quanitified the experssion using [Slamon Alevin method](https://salmon.readthedocs.io/en/latest/alevin.html).  

```{r}
txi.pbmc1 <- tximport("salmon.dir/pbmc_1k_v2_S1/alevin/quants_mat.gz", type="alevin")
d <- txi.pbmc1$counts

SYMBOL <- mapIds(EnsDb.Hsapiens.v75,keys=rownames(d),column="SYMBOL",keytype="GENEID",multiVals="first")
rownames(d) <- make.unique(ifelse(is.na(SYMBOL), rownames(d), SYMBOL))

pbmc1 <- CreateSeuratObject(counts = d, min.cells = 3, min.features = 200, project = "pbmc1")
pbmc1$cells <- "pbmc1"

txi.pbmc2 <- tximport("salmon.dir/pbmc_1k_v3_S1/alevin/quants_mat.gz", type="alevin")
d <- txi.pbmc2$counts

SYMBOL <- mapIds(EnsDb.Hsapiens.v75,keys=rownames(d),column="SYMBOL",keytype="GENEID",multiVals="first")
rownames(d) <- make.unique(ifelse(is.na(SYMBOL), rownames(d), SYMBOL))


pbmc2 <- CreateSeuratObject(counts = d, min.cells = 3, min.features = 200, project = "pbmc2")
pbmc2$cells <- "pbmc2"

```

# Assess QC metrics {.tabset .tabset-fade}


## pbmc1 features

In the example below, we visualize QC metrics, and use these to filter cells.

We filter cells that have unique feature counts over 2,500 or less than 200
We filter cells that have >5% mitochondrial counts

```{r}
pbmc1[["percent.mt"]] <- PercentageFeatureSet(pbmc1, pattern = "^MT-")
VlnPlot(pbmc1, c("nFeature_RNA", "nCount_RNA", "percent.mt"))
```

```{r}
plot1 <- FeatureScatter(pbmc1, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))
```

## pbmc2 features

```{r}
pbmc2[["percent.mt"]] <- PercentageFeatureSet(pbmc2, pattern = "^MT-")
VlnPlot(pbmc2, c("nFeature_RNA", "nCount_RNA", "percent.mt"))
```

```{r}
plot1 <- FeatureScatter(pbmc2, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))
```

# Filter the cells

```{r}
pbmc1 <- subset(pbmc1, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 25)
pbmc2 <- subset(pbmc2, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 25)

```

# Normalise the data

```{r}
pbmc1 <- NormalizeData(pbmc1)
pbmc2 <- NormalizeData(pbmc2)
```

# Identify highly variable genes

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

The procedure in Seurat3 is described in detail here, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the FindVariableFeatures function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.

```{r}
pbmc1 <- FindVariableFeatures(pbmc1, selection.method = "disp", nfeatures = 2000)
pbmc2 <- FindVariableFeatures(pbmc2, selection.method = "disp", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc1), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc1)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1
plot2
```

# Integrate both datasets 

Next I will use Seurat to integrate the datasets together into one object so I can compare clusters between the two datasets.

```{r}
integration <- FindIntegrationAnchors(object.list = list(pbmc1, pbmc2), dims=1:20)
to_integrate <- rownames(x = integration)
integration <- IntegrateData(anchorset = integration, dims = 1:20, features.to.integrate = to_integrate)

```

# Evaluate the clustering

Next I will use UMAP to visualise the clustering

First, we need to apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function:

Shifts the expression of each gene, so that the mean expression across cells is 0
Scales the expression of each gene, so that the variance across cells is 1
This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
The results of this are stored in integration[["RNA"]]@scale.data

```{r}
DefaultAssay(integration) <- "integrated"
integration <- ScaleData(integration)
```

# PCA

```{r}
integration <- RunPCA(integration, npcs = 30)
DimPlot(integration, reduction = "pca", split.by = "cells")
```

# UMAP

```{r}
integration <- RunUMAP(integration, reduction = "pca", dims = 1:20)
integration <- FindNeighbors(integration, reduction = "pca", dims = 1:20)
integration <- FindClusters(integration, resolution = 0.4)

p1 <- DimPlot(integration, reduction = "umap", group.by = "cells")
p2 <- DimPlot(integration, reduction = "umap", label = TRUE)
plot_grid(p1,p2)
```

# Clustering facet

```{r}
DimPlot(integration, reduction = "umap", split.by = "cells")
```

# Identify conserved cell type markers {.tabset .tabset-fade}

## Cluster 0

```{r}
DefaultAssay(integration) <- "RNA"
markers <- FindConservedMarkers(integration, ident.1 = 0, grouping.var = "cells")

datatable(markers)
```

## Cluster 1

```{r}
DefaultAssay(integration) <- "RNA"
markers <- FindConservedMarkers(integration, ident.1 = 1, grouping.var = "cells")

datatable(markers)
```

## Cluster 2

```{r}
DefaultAssay(integration) <- "RNA"
markers <- FindConservedMarkers(integration, ident.1 = 2, grouping.var = "cells")

datatable(markers)
```

## Cluster 3

```{r}
DefaultAssay(integration) <- "RNA"
markers <- FindConservedMarkers(integration, ident.1 = 3, grouping.var = "cells")

datatable(markers)
```

## Cluster 4

```{r}
DefaultAssay(integration) <- "RNA"
markers <- FindConservedMarkers(integration, ident.1 = 4, grouping.var = "cells")


datatable(markers)
```

### Conclusion from clustering analysis

Because of the low capture efficiency seen in the dataset, it has been quite challenging to identify bina fide "markers" of immune cell subsets. This suggeste that this experiment may need to be repeated with a better capture efficiency. 


# Differential between ctrl and j4 {.tabset .tabset-fade}

## Cluster 0

```{r}
integration$celltype.cells <- paste(Idents(integration), integration$cells, sep = "_")
integration$celltype <- Idents(integration)
Idents(integration) <- "celltype.cells"
zero.output <- FindMarkers(integration, ident.1 = "0_pbmc1", ident.2 = "0_pbmc2", verbose = FALSE)
datatable(zero.output)
```

## Cluster 1

```{r}
one.output <- FindMarkers(integration, ident.1 = "1_pbmc1", ident.2 = "1_pbmc2", verbose = FALSE)
datatable(one.output)
```



# Overlap between all outputs

```{r}
output <- intersect(rownames(zero.output),rownames(one.output))
output
```

# Feature plot

```{r}

FeaturePlot(integration, features = c("ATF3", "ATF4"), split.by = "cells",
            order = FALSE)
```
